filament/src/fg2/FrameGraph.h

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TNT_FILAMENT_WIP_FRAMEGRAPH_H
+#define TNT_FILAMENT_WIP_FRAMEGRAPH_H
+
+#include <backend/DriverEnums.h>
+#include <backend/Handle.h>
+
+namespace filament::fg2 {
+
+struct Texture {
+    struct Descriptor {
+        uint32_t width = 1;     // width of resource in pixel
+        uint32_t height = 1;    // height of resource in pixel
+        uint32_t depth = 1;     // # of images for 3D textures
+        uint8_t levels = 1;     // # of levels for textures
+        uint8_t samples = 0;    // 0=auto, 1=request not multisample, >1 only for NOT SAMPLEABLE
+        backend::SamplerType type = backend::SamplerType::SAMPLER_2D;     // texture target type
+        backend::TextureFormat format = backend::TextureFormat::RGBA8;    // resource internal format
+    };
+
+    enum class Usage {
+        SAMPLE,
+        UPLOAD
+    };
+
+    backend::Handle<backend::HwTexture> texture;
+};
+
+struct TextureSubresource {
+
+};
+
+
+/** A handle on a resource */
+class FrameGraphHandle;
+
+/** A typed handle on a resource */
+template<typename RESOURCE>
+class FrameGraphId<RESOURCE> : public FrameGraphHandle {
+public:
+    using FrameGraphHandle::FrameGraphHandle;
+    FrameGraphId() noexcept = default;
+    explicit FrameGraphId(FrameGraphHandle r) : FrameGraphHandle(r) { }
```

Comment:
Small idea: make this into a method called `cast<Foo>`, which would potentially help newcomers understand what the code is doing when it constructs one handle from another.  Otherwise they wonder if the resource is getting copied.

Reply:
ah... in fact this is not intended to be public I realize. I mean the user facing API should never use/see a non-typed framegraphid